Source/com/drew/imaging/png/PngChunkReader.java

```diff
@@ -86,7 +86,7 @@
             // Process the next chunk.
             int chunkDataLength = reader.getInt32();
 
-            if (chunkDataLength < 0)
+            if ((chunkDataLength < 0) || (chunkDataLength > reader.available()))
                 throw new PngProcessingException("PNG chunk length exceeds maximum");
 
             PngChunkType chunkType = new PngChunkType(reader.getBytes(4));
```

Comment:
For some reason I'm sure makes a lot of sense to GitHub, I can't place a comment any further down than here. I wanted to comment on line 96 below though.

I'm not saying that this will solve it, but I see one possibility to at least make this less likely to happen. As far as I can tell, `chunkData` is only used if `willStoreChunk` is `true`. It seems very wasteful to me to call `getBytes()` even if the data isn't to be kept, it will allocate memory and read information that will be just handed off to the GC. I would suggest something like this instead:
```java
byte[] chunkData;
if (willStoreChunk) {
  chunkData = reader.getBytes(chunkDataLength);
} else {
  chunkData = null // To satisfy the compiler
  reader.skip(chunkDataLength)
}
```

With that in place, I guess some sane maximum sizes could be deducted based on the chunk type. I'm no PNG expert, but I would assume that only the `IDAT` chunk could reasonably contain really large data, and I don't think that chunk is ever relevant for Metadata-extractor to read. For the chunk type of interest for Metadata-extractor, it should be possible to come up with some reasonable maximum value where everything above that would be rejected before calling `getBytes()` (which is where the allocation occurs)..?

Reply:
OK, this would be a nice workaround! In case of this specific file, the problematic chunk is not read, only skipped. So in the end `int chunkDataLength = reader.getInt32();` is negative in next loop iteration and `End of data reached.` is thrown. Nice!

I can see minor complication in the fact that `new PngChunkReader().extract(new StreamReader(inputStream), Collections.emptySet())` and `new PngChunkReader().extract(new StreamReader(inputStream), null)` can yield completely different result if your code snippet is used. Which is expected I guess, but should be mentioned in documentation/comments.

On the other hand, `PngMetadataReader.readMetadata(@NotNull InputStream inputStream)` has static set of `PngChunkType` defined, so no changes are required there :)

Reply:
I've implemented the changes you proposed, added a short comment regarding the chunk data reading and also included the corrupted file in the tests. Let me know if this solution is OK with you. Thanks!